File: sdk/python/tests/test_mcp_server.py
"""A minimal MCP server used as a subprocess for testing the SDK's adapter.

Supports both transports:

- ``python test_mcp_server.py``                     # stdio (default)
- ``python test_mcp_server.py http <port>``         # Streamable HTTP
"""

import sys

import anyio
from mcp import types
from mcp.server.fastmcp import FastMCP
from mcp.server.lowlevel import Server
from mcp.server.stdio import stdio_server
from mcp.types import ToolAnnotations

server = FastMCP("test")


@server.tool(annotations=ToolAnnotations(readOnlyHint=True))
def greet(name: str) -> str:
    """Greet someone by name."""
    return f"Hello, {name}!"


@server.tool()
def add(a: int, b: int) -> str:
    """Add two numbers."""
    return str(a + b)


@server.resource("test://item/{item_id}")
def get_item(item_id: str) -> str:
    """Get an item by ID."""
    return f"Item {item_id}"


@server.prompt()
def summarize(text: str) -> str:
    """Summarize the given text."""
    return f"Please summarize: {text}"


tools_only_server = Server("tools-only")


@tools_only_server.list_tools()
async def list_tools() -> list[types.Tool]:
    return [
        types.Tool(
            name="greet",
            description="Greet someone",
            inputSchema={
                "type": "object",
                "properties": {"name": {"type": "string"}},
                "required": ["name"],
            },
            annotations=ToolAnnotations(readOnlyHint=True),
        )
    ]


@tools_only_server.call_tool()
async def call_tool(name: str, arguments: dict[str, object]) -> list[types.TextContent]:
    return [types.TextContent(type="text", text=f"Hello, {arguments['name']}!")]


async def run_tools_only_server() -> None:
    async with stdio_server() as (read_stream, write_stream):
        await tools_only_server.run(
            read_stream,
            write_stream,
            tools_only_server.create_initialization_options(),
        )


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "http":
        port = int(sys.argv[2]) if len(sys.argv) > 2 else 8765
        server.settings.host = "127.0.0.1"
        server.settings.port = port
        server.run(transport="streamable-http")
    elif len(sys.argv) > 1 and sys.argv[1] == "tools-only":
        anyio.run(run_tools_only_server)
    else:
        server.run(transport="stdio")

File: sdk/typescript/tests/fixtures/test-mcp-server.mjs
// Minimal MCP server used as a subprocess for testing the SDK's adapter.
// Supports both transports:
//   node test-mcp-server.mjs                  # stdio (default)
//   node test-mcp-server.mjs http <port>      # Streamable HTTP
import { McpServer, ResourceTemplate } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { z } from 'zod';

function buildServer() {
  const server = new McpServer({ name: 'test', version: '1.0.0' });

  server.registerTool(
    'greet',
    {
      description: 'Greet someone by name',
      annotations: { readOnlyHint: true },
      inputSchema: { name: z.string().describe('Person to greet') },
    },
    async (args) => ({
      content: [{ type: 'text', text: `Hello, ${args.name}!` }],
    })
  );

  server.tool(
    'add',
    'Add two numbers',
    { a: z.number(), b: z.number() },
    async (args) => ({
      content: [{ type: 'text', text: String(args.a + args.b) }],
    })
  );

  server.resource(
    'item',
    new ResourceTemplate('test://item/{item_id}', { list: undefined }),
    async (uri, args) => ({
      contents: [
        {
          uri: uri.href,
          text: `Item ${args.item_id}`,
          mimeType: 'text/plain',
        },
      ],
    })
  );

  server.prompt(
    'summarize',
    'Summarize the given text',
    { text: z.string() },
    async (args) => ({
      messages: [
        {
          role: 'user',
          content: { type: 'text', text: `Please summarize: ${args.text}` },
        },
      ],
    })
  );

  return server;
}

async function runStdio() {
  const server = buildServer();
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

async function runHttp(port) {
  const { StreamableHTTPServerTransport } = await import(
    '@modelcontextprotocol/sdk/server/streamableHttp.js'
  );
  const http = await import('node:http');

  const httpServer = http.createServer(async (req, res) => {
    if (!req.url || !req.url.startsWith('/mcp')) {
      res.statusCode = 404;
      res.end('not found');
      return;
    }
    const chunks = [];
    for await (const chunk of req) {
      chunks.push(chunk);
    }
    const bodyRaw = Buffer.concat(chunks).toString('utf8');
    const body = bodyRaw ? JSON.parse(bodyRaw) : undefined;

    // Stateless: spin up a fresh server + transport per request. Simpler than
    // tracking sessions across requests for a test fixture.
    const transport = new StreamableHTTPServerTransport({
      sessionIdGenerator: undefined,
    });
    const server = buildServer();
    await server.connect(transport);
    res.on('close', () => {
      transport.close();
      server.close();
    });
    await transport.handleRequest(req, res, body);
  });

  await new Promise((resolve) => httpServer.listen(port, '127.0.0.1', resolve));
  process.stdin.resume();
}

const mode = process.argv[2];
if (mode === 'http') {
  const port = Number(process.argv[3] ?? 8765);
  runHttp(port).catch((err) => {
    console.error('http fixture error:', err);
    process.exit(1);
  });
} else {
  runStdio().catch((err) => {
    console.error('stdio fixture error:', err);
    process.exit(1);
  });
}
